### ISSUE TITLE
Batchnorm Scaling Factor is Clamped Near Zero

### ISSUE THREAD

**Issue by whuffman** [martinarjovsky/WassersteinGAN#18]
I believe that the parameter clamping also reduces the batchnorm scaling factor to near zero, when (as far as I understand) it should stay near 1 ([where it was initialized](https://github.com/martinarjovsky/WassersteinGAN/blob/7eee65dc9cd08414dfcd5718b5cbe7b4c0360495/main.py#L98-L105)).

[In line 170 of main.py:](https://github.com/martinarjovsky/WassersteinGAN/blob/7eee65dc9cd08414dfcd5718b5cbe7b4c0360495/main.py#L170-L172)
```
# clamp parameters to a cube
for p in netD.parameters():
    p.data.clamp_(opt.clamp_lower, opt.clamp_upper)
```

**Comment by NickShahML** [comment]
This makes alot of sense to me. I'm going to take out clamping for batch norm scaling and take this part out and see if it leads to any improvements.

**Comment by rafaelvalle** [comment]
@NickShahML please share your results once you have them!

**Comment by ppwwyyxx** [comment]
I've tried it before. The loss becomes several magnitudes higher, and results are not as good.
Mentioned it a month ago on [reddit](https://www.reddit.com/r/MachineLearning/comments/5qxoaz/r_170107875_wasserstein_gan/dddwr1k/?utm_content=permalink&utm_medium=front&utm_source=reddit&utm_name=MachineLearning).

**Comment by NickShahML** [comment]
I'm getting exactly what @ppwwyyxx reported. The results are simply chaotic and far higher of loss.

**Comment by whuffman** [comment]
Interesting!  Good to know - I have been trying to get a wgan working on a separate codebase, and I'm running into trouble.  I'll try putting my batchnorm scale around the same magnitude as my weight clamping.

Anyway, if you make a small change to the weights_init function to have the batchnorm scale factors start at 0.01, instead of near 1, it may help others reading through the code!

**Comment by ppwwyyxx** [comment]
Actually, I also found that as long as you did the clipping (and clip BN as well), it will have some of the nice properties of WGAN, even just using the original GAN loss. For example, it can train without requiring BatchNorm in generator.

**Comment by martinarjovsky** [comment]
If your network is linear with no biases scaling the BN parameter by a constant c would multiply the output of the network by c. Since the discriminator wants to make the loss as large as possible, this c would go to infinity very quickly, that's why it should be clamped (same as with most other parameters in the network).
